[BUG] non-deterministic compiled SQLExecPlugin.class with scala 2.13 deployment

**Describe the bug**
We saw it failed in our CI `not bitwise-identical across shims` constantly.

```
[2023-10-29T11:34:31.265Z] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:3.1.0:run (create-parallel-world) on project rapids-4-spark_2.13: An Ant BuildException has occured: The following error occurred while executing this line:
[2023-10-29T11:34:31.265Z] [ERROR] /home/jenkins/agent/workspace/jenkins-rapids_scala213_nightly-dev-github-4/dist/maven-antrun/build-parallel-worlds.xml:124: exec binary-dedupe.sh failed, exit code is 255, error msg is com/nvidia/spark/rapids/SQLExecPlugin.class is not bitwise-identical across shims
```


after comparing the generated SQLExecPlugin.class in different spark shims, I found the only diff is that
mvn install
```
public <A$> Function1<A$, BoxedUnit> compose(final Function1<A$, SparkSessionExtensions> g) {
        return Function1.compose$(this, g);
    }

    public <A$> Function1<SparkSessionExtensions, A$> andThen(final Function1<BoxedUnit, A$> g) {
        return Function1.andThen$(this, g);
    }
```
after following mvn deploy:
```
    public <A> Function1<A, BoxedUnit> compose(final Function1<A, SparkSessionExtensions> g) {
        return Function1.compose$(this, g);
    }

    public <A> Function1<SparkSessionExtensions, A> andThen(final Function1<BoxedUnit, A> g) {
        return Function1.andThen$(this, g);
    }
```

`A$ vs A`

**Steps/Code to reproduce bug**
run with internal `rapids_scala213_nightly-dev-github` pipeline (dup a pipeline for testing)
**UPDATE:** local repro steps https://github.com/NVIDIA/spark-rapids/issues/9571#issuecomment-1784615813

**Expected behavior**
After build SQLExecPlugin.class should always have the identical sha value in different shims, 
or move SQLExecPlugin.class out of check

--- a/sql-plugin/src/main/scala-2.12/com/nvidia/spark/rapids/RapidsStack.scala
+++ b/sql-plugin/src/main/scala-2.12/com/nvidia/spark/rapids/RapidsStack.scala
@@ -16,6 +16,37 @@
 
 package com.nvidia.spark.rapids
 
-import scala.collection.mutable.Stack
+import scala.collection.mutable.ArrayStack
 
-class ScalaStack[T] extends Stack[T]
+class RapidsStack[T] {
+  private val innerStack = new ArrayStack[T]()
+
+  def push(elem1: T): RapidsStack[T] = {
+    innerStack.push(elem1)
+    this
+  }
+
+  def pop(): T = {
+    innerStack.pop()
+  }
+
+  def isEmpty: Boolean = {
+    innerStack.isEmpty
+  }
+
+  def nonEmpty: Boolean = {
+    innerStack.nonEmpty
+  }
+
+  def size(): Int = {
+    innerStack.size
+  }
+
+  def toSeq(): Seq[T] = {
+    innerStack.toSeq
+  }
+
+  def clear(): Unit = {
+    innerStack.clear()
+  }
+}

--- a/sql-plugin/src/main/scala-2.12/com/nvidia/spark/rapids/RapidsStack.scala
+++ b/sql-plugin/src/main/scala-2.12/com/nvidia/spark/rapids/RapidsStack.scala
@@ -19,11 +19,12 @@ package com.nvidia.spark.rapids
 import scala.collection.mutable.ArrayStack
 
 class RapidsStack[T] extends Proxy {
-  override def self = new ArrayStack[T]()
+  private val stack = new ArrayStack[T]()
 
-  def push(elem1: T): RapidsStack[T] = {
+  override def self = stack
+
+  def push(elem1: T): Unit = {
     self.push(elem1)
-    this
   }
 
   def pop(): T = {
@@ -38,11 +39,11 @@ class RapidsStack[T] extends Proxy {
     self.nonEmpty
   }
 
-  def size(): Int = {
+  def size: Int = {
     self.size
   }
 
-  def toSeq(): Seq[T] = {
+  def toSeq: Seq[T] = {
     self.toSeq
   }
 

--- a/sql-plugin/src/main/scala-2.12/com/nvidia/spark/rapids/RapidsStack.scala
+++ b/sql-plugin/src/main/scala-2.12/com/nvidia/spark/rapids/RapidsStack.scala
@@ -18,35 +18,35 @@ package com.nvidia.spark.rapids
 
 import scala.collection.mutable.ArrayStack
 
-class RapidsStack[T] {
-  private val innerStack = new ArrayStack[T]()
+class RapidsStack[T] extends Proxy {
+  override def self = new ArrayStack[T]()
 
   def push(elem1: T): RapidsStack[T] = {
-    innerStack.push(elem1)
+    self.push(elem1)
     this
   }
 
   def pop(): T = {
-    innerStack.pop()
+    self.pop()
   }
 
   def isEmpty: Boolean = {
-    innerStack.isEmpty
+    self.isEmpty
   }
 
   def nonEmpty: Boolean = {
-    innerStack.nonEmpty
+    self.nonEmpty
   }
 
   def size(): Int = {
-    innerStack.size
+    self.size
   }
 
   def toSeq(): Seq[T] = {
-    innerStack.toSeq
+    self.toSeq
   }
 
   def clear(): Unit = {
-    innerStack.clear()
+    self.clear()
   }
 }

--- a/sql-plugin/src/main/scala-2.13/com/nvidia/spark/rapids/RapidsStack.scala
+++ b/sql-plugin/src/main/scala-2.13/com/nvidia/spark/rapids/RapidsStack.scala
@@ -18,35 +18,35 @@ package com.nvidia.spark.rapids
 
 import scala.collection.mutable.Stack
 
-class RapidsStack[T] {
-  private val innerStack = new Stack[T]()
+class RapidsStack[T] extends scala.Proxy {
+  override def self = new Stack[T]()
 
   def push(elem1: T): RapidsStack[T] = {
-    innerStack.push(elem1)
+    self.push(elem1)
     this
   }
 
   def pop(): T = {
-    innerStack.pop()
+    self.pop()
   }
 
   def isEmpty: Boolean = {
-    innerStack.isEmpty
+    self.isEmpty
   }
 
   def nonEmpty: Boolean = {
-    innerStack.nonEmpty
+    self.nonEmpty
   }
 
   def size(): Int = {
-    innerStack.size
+    self.size
   }
 
   def toSeq(): Seq[T] = {
-    innerStack.toSeq
+    self.toSeq
   }
 
   def clear(): Unit = {
-    innerStack.clear()
+    self.clear()
   }
 }

--- a/sql-plugin/src/main/scala-2.13/com/nvidia/spark/rapids/RapidsStack.scala
+++ b/sql-plugin/src/main/scala-2.13/com/nvidia/spark/rapids/RapidsStack.scala
@@ -16,6 +16,37 @@
 
 package com.nvidia.spark.rapids
 
-import scala.collection.mutable.ArrayStack
+import scala.collection.mutable.Stack
 
-class ScalaStack[T] extends ArrayStack[T]
+class RapidsStack[T] {
+  private val innerStack = new Stack[T]()
+
+  def push(elem1: T): RapidsStack[T] = {
+    innerStack.push(elem1)
+    this
+  }
+
+  def pop(): T = {
+    innerStack.pop()
+  }
+
+  def isEmpty: Boolean = {
+    innerStack.isEmpty
+  }
+
+  def nonEmpty: Boolean = {
+    innerStack.nonEmpty
+  }
+
+  def size(): Int = {
+    innerStack.size
+  }
+
+  def toSeq(): Seq[T] = {
+    innerStack.toSeq
+  }
+
+  def clear(): Unit = {
+    innerStack.clear()
+  }
+}

--- a/sql-plugin/src/main/scala-2.13/com/nvidia/spark/rapids/RapidsStack.scala
+++ b/sql-plugin/src/main/scala-2.13/com/nvidia/spark/rapids/RapidsStack.scala
@@ -19,7 +19,8 @@ package com.nvidia.spark.rapids
 import scala.collection.mutable.Stack
 
 class RapidsStack[T] extends Proxy {
-  override def self = new Stack[T]()
+  private val stack = new Stack[T]()
+  override def self = stack
 
   def push(elem1: T): RapidsStack[T] = {
     self.push(elem1)

--- a/sql-plugin/src/main/scala-2.13/com/nvidia/spark/rapids/RapidsStack.scala
+++ b/sql-plugin/src/main/scala-2.13/com/nvidia/spark/rapids/RapidsStack.scala
@@ -18,7 +18,7 @@ package com.nvidia.spark.rapids
 
 import scala.collection.mutable.Stack
 
-class RapidsStack[T] extends scala.Proxy {
+class RapidsStack[T] extends Proxy {
   override def self = new Stack[T]()
 
   def push(elem1: T): RapidsStack[T] = {

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala
@@ -450,8 +450,8 @@ case class GpuOutOfCoreSortIterator(
     while (!pending.isEmpty && sortedSize < targetSize) {
       // Keep going until we have enough data to return
       var bytesLeftToFetch = targetSize
-      val pendingSort = new ScalaStack[SpillableColumnarBatch]()
-      closeOnExcept(pendingSort) { _ =>
+      val pendingSort = new RapidsStack[SpillableColumnarBatch]()
+      closeOnExcept(pendingSort.toSeq) { _ =>
         while (!pending.isEmpty &&
             (bytesLeftToFetch - pending.peek().buffer.sizeInBytes >= 0 || pendingSort.isEmpty)) {
           val buffer = pending.poll().buffer

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala
@@ -242,9 +242,9 @@ class GpuSorter(
    * @return the sorted data.
    */
   final def mergeSortAndCloseWithRetry(
-      spillableBatches: ScalaStack[SpillableColumnarBatch],
+      spillableBatches: RapidsStack[SpillableColumnarBatch],
       sortTime: GpuMetric): SpillableColumnarBatch = {
-    closeOnExcept(spillableBatches) { _ =>
+    closeOnExcept(spillableBatches.toSeq) { _ =>
       assert(spillableBatches.nonEmpty)
     }
     withResource(new NvtxWithMetrics("merge sort", NvtxColor.DARK_GREEN, sortTime)) { _ =>
@@ -277,9 +277,9 @@ class GpuSorter(
             }
           }
         } else {
-          closeOnExcept(spillableBatches) { _ =>
-            val batchesToMerge = new ScalaStack[SpillableColumnarBatch]()
-            closeOnExcept(batchesToMerge) { _ =>
+          closeOnExcept(spillableBatches.toSeq) { _ =>
+            val batchesToMerge = new RapidsStack[SpillableColumnarBatch]()
+            closeOnExcept(batchesToMerge.toSeq) { _ =>
               while (spillableBatches.nonEmpty || batchesToMerge.size > 1) {
                 // pop a spillable batch if there is one, and add it to `batchesToMerge`.
                 if (spillableBatches.nonEmpty) {
@@ -299,7 +299,7 @@ class GpuSorter(
 
                   // we no longer care about the old batches, we closed them
                   closeOnExcept(merged) { _ =>
-                    batchesToMerge.safeClose()
+                    batchesToMerge.toSeq.safeClose()
                     batchesToMerge.clear()
                   }
 

